### ISSUE TITLE
Circular files import causing "Unexpected Name "undefined""

### ISSUE THREAD

**Issue by pwortham** [apollographql/graphql-tag#262]
Hi, 

I have a **_Syntax Error: Unexpected Name "undefined"_** error when I have circular imports between fragments files.
I don't have circular dependencies on my fragment, just with the files import.

For example:

 ```
// fragmentA.ts

import gql from "graphql-tag";
import {fragmentB} from "./fragmentB";

export const fragmentA = gql`fragment aFields on A {
    ...bFields
}
${fragmentB}`;
```

```
// fragmentB.ts

import gql from "graphql-tag";
import {fragmentA} from "./fragmentA";

export const fragmentB = gql`fragment bFields on B {
    id
}
`;

export const fragmentC = gql`fragment cFields on C {
    ...aFields
}
${fragmentA}`;

```
I have now an error on fragmentA cause of fragmentC.

I reproduce this example on this repo: https://github.com/pwortham/graphql-fragments

**Comment by jcardali** [comment]
Will this be addressed?

**Comment by zingerj** [comment]
I also ran into this today. Is there a fix planned?

**Comment by sean6bucks** [comment]
Ran into this while trying to use Fragments with react Context api which is causing "circular" imports while trying to run Jest tests

**Comment by brussev** [comment]
Did someone found any workaround to the problem?

**Comment by cgonzalezsan** [comment]
+1

**Comment by yoshig** [comment]
+1, still having issues with this

**Comment by philetos** [comment]
I ran into the same error, however all my fragments where defined in one file. 

My solution was to make sure all "dependent" fragments were defined first in the file (top to bottom).  Not sure if this could resolve all use cases...but it helped me :]

**Comment by batamire** [comment]
@philetos solution plus i had to import the file directly, didn't work through `directory/index.js`... very weird.

A terribly cryptic error, a better message would save countless hours around the globe 😉

**Comment by PhilReinking** [comment]
I just also came across this issue. Is this something that is going to be fixed anytime?

**Comment by Athelian** [comment]
> @philetos solution plus i had to import the file directly, didn't work through `directory/index.js`... very weird.
> 
> A terribly cryptic error, a better message would save countless hours around the globe 😉

I am also having to import things directly rather than through collated `directory/index.js`s. Did you find a better solution for this? Removing each `index.js` of each directory means I lose a lot of organization.

**Comment by camilo-mujica** [comment]
Same problem here, any updates on this one?

**Comment by phryneas** [comment]
This is unfortunately nothing we can address. 

JavaScript evaluates each file in full from top to bottom before evaluating another file. 
So in this case, `fragmentA.ts` cannot be evaluated because it relies on `fragmentB.ts` - but `fragmentB.ts` cannot be evaluated because it relies on `fragmentA.ts`.

There is the case where a bundler could statically analyze the code, and when merging it into a full file, reorder it - to be like

```js
export const fragmentB = ...
export const fragmentA = ...
export const fragmentC = ...
```

But in most situations, this will not happen, as this would execute possible side effects out of order, and I don't know of any bundler that would actually take that risk.

There are other kinds of circles that *can* be resolved by a bundler, by making sure that file A always executed before file B - e.g. if all references in file B to file A are within a function that is not immediately executed - but the example given is not one of those.

---

## To sum it up:  

The code presented here is just "impossible" to be evaluated by a normal JavaScript engine that just executes code top-to-bottom.  
There is nothing we can do about it, but you can fix it on your side:  
Break the circle by adding one more file to the equation having `fragmentB` and `fragmentC` in different files.

---

As there is nothing we can do about it (really, it's absolutely impossible!), I'm going to close this issue.
